examples/cfq/syntax-based/asg_test.py

```diff
@@ -0,0 +1,111 @@
+from absl.testing import absltest
+from absl.testing import parameterized
+from asg import apply_rule_act, generate_act, generate_action_sequence
+from grammar import Grammar, GRAMMAR_STR
+
+class AsgTest(parameterized.TestCase):
+
+
+  def test_generate_action_sequence_muliple(self):
```

Comment:
Currently you aren't testing the actual string output but you are using `apply_rule_act` and `generate_act`, which makes it difficult to verify if we are actually outputting what we want. Usually in tests you want to use as few dependencies on outside data as possible, since those could contain a bug that we may miss now. 

Similarly, I think the better way of testing this is to create a (simple) `GRAMMAR_STR` in this file, since currently you are using one from another file, which also could lead to unexpected results if someone decided to change that string. (I also makes it difficult for me to understand what is going on, since GRAMMAR_STR is somewhere else.) Then you could create another simple query and you could test that in isolation without having any other dependencies.

But if that is too much work, then a good first step is to simply put the actual strings in `expected_action_sequence`.

Reply:
That makes sense. It was easier to write the tests with apply_rule_act and generate_act, and maybe a bit easier to read as well. I will think a bit about it, I am not sure on the representation of the rules yet, so I wanted it to be easily modifiable later on.

About having the grammar string in the test file: the grammar and the code for generating the actions are tightly linked now (that is, if the grammar would be changed, the code has to be changed as well). This means that as long as all the rules are used in the tests (which is the case), I would just define it in the test file again. Alternatively, I could define a shorter version for each test, but not much shorter and I am not sure that would make things more clear.